### PR TITLE
More /vg/lighting tweaks

### PR DESCRIPTION
**Pull request by MrsTonedOne**
Rather than remove the light from all tiles, then re-add the light to all tiles, we just go thru each tile and diff the light level from the last value we added to it. (since this is tracked)
This cut the proc calls for updating light sources on movements in half.

Lighting objects now default to full brite rather than full dark so shuttles aren't as immersion breaking when they transit.

Made lighting more aggressive about clearing empty lists. (Memory usage went down ~80mb)

Break all lights followed by fix all lights went down from 3.5s cpu time to 2.4s cpu time.

Because this is a decent sized change in how lighting_source updates operate, its hard to get a true apples to apples comparison by comparing two procs like i normally would.

:cl:
tweak: Lighting now defaults to fully bright until the first update tick for that tile. This makes shuttle movements less immersion breaking.
experiment: Rather than remove the light source from the system and then re-adding it on light movements, the system now calculates and applies the difference to each tile. This should speed up lighting updates.
/:cl:

@PJB3005 i merged smart_vis_update and apply_lum and check() into one proc by making APPLY_CORNER() diff the old value (if any) from the new value.

@Incoming5643 this should fix tiles that started out dark not getting the dark icon state

### CODE REVIEW

File: code/modules/lighting/lighting_object.dm [hunk 1]
```diff
@@ -69,10 +68,17 @@ GLOBAL_LIST_EMPTY(all_lighting_objects) // Global list of lighting objects.
...
-	var/datum/lighting_corner/cb  = T.corners[4] || dummy_lighting_corner
-	var/datum/lighting_corner/ca  = T.corners[1] || dummy_lighting_corner
+
+	var/list/corners = T.corners
+	var/datum/lighting_corner/cr = dummy_lighting_corner
+	var/datum/lighting_corner/cg = dummy_lighting_corner
+	var/datum/lighting_corner/cb = dummy_lighting_corner
+	var/datum/lighting_corner/ca = dummy_lighting_corner
+	if (corners) //done this way for speed
+		cr = T.corners[3] || dummy_lighting_corner
```
Comment (RemieRichards) [inline code/modules/lighting/lighting_object.dm]: could you not then accessing these from ```corners```, the local should be faster

Comment (MrStonedOne) [inline code/modules/lighting/lighting_object.dm]: oh ya

File: code/modules/lighting/lighting_source.dm [hunk 2]
```diff
@@ -164,48 +110,47 @@
...
-		. * applied_lum_r,           \
-		. * applied_lum_g,           \
-		. * applied_lum_b            \
+
+#define APPLY_CORNER(C)                      \
+	. = LUM_FALLOFF(C, pixel_turf);          \
+	. *= light_power;                        \
+	var/OLD = effect_str[C];                 \
+                                             \
+	effect_str[C] = .;                       \
```
Comment (MrPerson) [inline code/modules/lighting/lighting_source.dm]: you could do this on one line with effect_str[C] = . = LUM_FALLOFF(C, pixel_turf) * light_power although that might be a too much

**Comment by ChangelingRain** [comment]
Yeah space doesn't light walls, it only emits light next to open turfs

**Comment by ChangelingRain** [comment]
From that issue, I can say emitter beams also cause that.

**Comment by octareenroon91** [comment]
See those dark spots? I only grew and removed glowshrooms. Zero shadowshrooms.

There was also a weird shadowy spot in the arrival shuttle before this round.

screenshot_2017-05-06_22-32-00
screenshot_2017-05-06_22-35-54

Values saturate but subtraction doesn't know.

File: code/modules/lighting/lighting_source.dm [hunk 3]
```diff
@@ -164,48 +110,47 @@
...
+                                             \
+	C.update_lumcount                        \
+	(                                        \
+		(. * lum_r) - (OLD * applied_lum_r), \
+		(. * lum_g) - (OLD * applied_lum_g), \
+		(. * lum_g) - (OLD * applied_lum_g)  \
```
Comment (Lohikar) [inline code/modules/lighting/lighting_source.dm]: Shouldn't this last line be `(. * lum_b) - (OLD * applied_lum_b)`?

Comment (MrStonedOne) [inline code/modules/lighting/lighting_source.dm]: yes

**Comment by octareenroon91** [comment]
screenshot_2017-05-06_23-23-07

**Comment by SpinnerMaster** [comment]
Can confirm arrivals shuttle lighting is fucked.

image

**Comment by MrStonedOne** [comment]
all the issues seem to be fixed.

**Comment by AnturK** [comment]
Still happening ? Or was the testmerge not updated ?

**Comment by MrStonedOne** [comment]
hmmm

**Comment by MrStonedOne** [comment]
the test merge was fixed, so this would be it still happening.

**Comment by octareenroon91** [comment]
>Rather than remove the light from all tiles, then re-add the light to all tiles, we just go thru each tile and diff the light level from the last value we added to it. (since this is tracked)

This change is sensitive to inconsistency in light creation/removal.

Also I'm assuming that light levels are in boundaries (-infinity, 1] so lights that don't clean up after themselves in the sanest ways will pull a turf's light level down for the rest of the round.

Consider adding occasional full recalculations.

**Comment by PJB3005** [comment]
@octareenroon91 light levels have no bounds other than the limits of a single precision float. They get capped in the calculation for the strength of a lighting overlay and `turf.get_lumcount()` though.

**Comment by MrStonedOne** [comment]
>Consider adding occasional full recalculations.

So, I'm going to point out the irony that the current bug antur mentioned, is actually from a full recalculation that happens when turfs change.

**Comment by optimumtact** [comment]
those bugs look hilarious haha, we should have some kind of light that can achieve that effect.

**Comment by ChangelingRain** [comment]
Still SOME issues.

**Comment by Razharas** [comment]
Cut and build instead of readjusting is a method plaguing our codebase in all systems
We need more of this

**Comment by PJB3005** [comment]
I was totally thinking of being more efficient with the moving of light sources, but it just gave me a giant headache to think about and all the edge cases and I gave up.

Vis updates do only update what needs updating already though.

**Pull request merged**